### PR TITLE
Code review

### LINKED ISSUE #5: Review process: delete branches after merge
Assuming we have a page on the review process, we should mention that we our practice is to delete branches (from github) after merging.

### PR DESCRIPTION
**Pull request by justinabrahms**
Some notes around code review. This was pulled from @marionzualo's gist at https://gist.github.com/marionzualo/82fbf4e3d20ef253d3ef.

I think this captures our code review process pretty well. I'd love to hear the broader team's thoughts on this.

Fixes #5

### CODE REVIEW

File: code_review.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+* Is all the code easily understood? Is code unnecessarily complex ?
+* Can variable/method names be improved?
+* Are classes/files/methods becoming too large? Should they be split into smaller pieces?
+* Does it conform to the agreed coding conventions?
+* Is there any redundant or duplicate code? Is the code as modular as possible?
+* Could the design of the solution be improved (e.g. single responsibility principle, open/closed principle) ?
+* Is the code testable? (i.e. don't add too many or hide dependencies, unable to initialize objects, test frameworks can use methods, do tests exist and are they comprehensive?)
+* Does the code work? Does it perform its intended function, the logic is correct etc.
+* Things related to your area of expertiese
```
Comment (pdpinch) [inline code_review.md]: s/expertiese/expertise/

File: pr-template.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#### What's this PR do?
+
+#### Where should the reviewer start?
+
+#### How should this be manually tested?
+
+#### Any background context you want to provide?
+
+#### What are the relevant tickets?
```
Comment (pdpinch) [inline pr-template.md]: I'd like to move this up the list, maybe even to the first position. 

We should also use the github keywords for [closing issues via commit messages](https://help.github.com/articles/closing-issues-via-commit-messages/)

File: code_review.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+* Try to respond to every comment.
+* Wait to merge the branch until someone has :+1:'d your PR and Continuous Integration
+tells you the test suite is green in the branch.
+* Merge once you feel confident in the code and its impact on the project.
+* Delete the backing branch.
+
+# Reviewing Code
+* Code review often and for short sessions. The effectiveness of code review drops after around an hour. Set aside time throughout your day to coincide with breaks in order to not disrupt your flow.
+* Review at least part of the code, even if you can't do all of it.
+* Review the right things, let tools to do the rest. Don't focus on reviewing things that can be reviewed by tools (e.g. Rubocop, jsLint, sass lint, etc). If the tools don't provide good feedback, change the config or improve the tools.
```
Comment (pdpinch) [inline code_review.md]: Is it worth updating this list of tools with the ones we do use?

File: code_review.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+* Provide context about what the change is and why is it important - Write informative commit messages [4] and pull request descriptions.
+* Use the pull request description template [3]. You only need to answer the relevant questions. Pro tip: use the bookmarklet [11] or chrome extension [12] for pull request description templates.
+* Annotate your code reviews. Add comments anywhere that you feel will help the reviewer understand your code.
+* Be grateful for the reviewer's suggestions. ("Good call. I'll make that
+change.")
+* Don't take it personally. The review is about the code, not you.
+* Explain why the code exists. ("It's like that because of these reasons. Would
+it be more clear if I rename this class/file/method/variable?")
+* Seek to understand the reviewer's perspective.
+* Try to respond to every comment.
```
Comment (pdpinch) [inline code_review.md]: I'd like to be more forceful about this, perhaps;

```
Respond to every comment. If you can't or won't change something, let the reviewers know why.
```

File: code_review.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+* Wait to merge the branch until someone has :+1:'d your PR and Continuous Integration
+tells you the test suite is green in the branch.
+* Merge once you feel confident in the code and its impact on the project.
+* Delete the backing branch.
+
+# Reviewing Code
+* Code review often and for short sessions. The effectiveness of code review drops after around an hour. Set aside time throughout your day to coincide with breaks in order to not disrupt your flow.
+* Review at least part of the code, even if you can't do all of it.
```
Comment (pdpinch) [inline code_review.md]: One thing I worry about are partial reviews -- that either block merges, or imply a thumbs up. 

If you can't review everything, how do you signal the author that you'll be back later, or that you need someone else to complete the review?

File: code_review.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+Some suggestions of things to look for when reviewing code:
+
+* Is all the code easily understood? Is code unnecessarily complex ?
+* Can variable/method names be improved?
+* Are classes/files/methods becoming too large? Should they be split into smaller pieces?
+* Does it conform to the agreed coding conventions?
+* Is there any redundant or duplicate code? Is the code as modular as possible?
+* Could the design of the solution be improved (e.g. single responsibility principle, open/closed principle) ?
+* Is the code testable? (i.e. don't add too many or hide dependencies, unable to initialize objects, test frameworks can use methods, do tests exist and are they comprehensive?)
+* Does the code work? Does it perform its intended function, the logic is correct etc.
```
Comment (pdpinch) [inline code_review.md]: This seems like it should be higher up the list. Or the very last bullet.

File: code_review.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+Some suggestions of things to look for when reviewing code:
+
+* Is all the code easily understood? Is code unnecessarily complex ?
+* Can variable/method names be improved?
+* Are classes/files/methods becoming too large? Should they be split into smaller pieces?
+* Does it conform to the agreed coding conventions?
+* Is there any redundant or duplicate code? Is the code as modular as possible?
+* Could the design of the solution be improved (e.g. single responsibility principle, open/closed principle) ?
+* Is the code testable? (i.e. don't add too many or hide dependencies, unable to initialize objects, test frameworks can use methods, do tests exist and are they comprehensive?)
```
Comment (pdpinch) [inline code_review.md]: This should be more rigorous:

``` 
Are there tests? If not, why not? Is there an issue open to add them later?
```

The parenthetical is also confusing, if it's still needed.

File: code_review.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+* [5] [Am I My Code?](http://mfeckie.github.io/Am-I-My-Code/) - this article explores some strategies for giving feedback when reviewing code.
+* [6] [Code review at the Dropbox iOS team](http://www.objc.io/issue-22/dropbox.html) - the article goes over their whole process of review. It has some interesting ideas.
+* [7] [Code Review Checklist](http://blog.fogcreek.com/increase-defect-detection-with-our-code-review-checklist-example/) - from Fog Creek Software
+* [8] [Effective code review tips](http://blog.fogcreek.com/effective-code-reviews-9-tips-from-a-converted-skeptic/) - from Fog Creek Software
+* [9] [Code Review Best Practices](http://kevinlondon.com/2015/05/05/code-review-best-practices.html) - really good article on code review best practices
+* [10] [11 proven practices to improve code reviews](http://www.ibm.com/developerworks/rational/library/11-proven-practices-for-peer-review/) - nice article backed by data from studies.
+* [11] [Pull request template bookmarklet](https://quickleft.com/blog/pull-request-template-bookmarklet/)
+* [12] [Pull request template chrome extension](https://github.com/sprintly/pull-request-template-chrome-extension)
```
Comment (pdpinch) [inline code_review.md]: That's a lot of references. @justinabrahms do you think they're all as useful as the original author did?

**Comment by pdpinch** [comment]
Since this has been pulled from someone else's gist, should we acknowledge that in the code, and not just the PR description? (And is there any reason to be concerned about copyright?).

**Comment by justinabrahms** [comment]
I don't think there are issues with copyright. https://twitter.com/marionzualo/status/677126919968718849

File: code_review.md (hunk 8 above)
Comment (justinabrahms) [inline code_review.md]: They're all pretty good, imo.

File: code_review.md (hunk 1 above)
Comment (justinabrahms) [inline code_review.md]: Done.

File: pr-template.md (hunk 2 above)
Comment (justinabrahms) [inline pr-template.md]: Done.

File: code_review.md (hunk 3 above)
Comment (justinabrahms) [inline code_review.md]: Done.

File: code_review.md (hunk 4 above)
Comment (justinabrahms) [inline code_review.md]: Done.

File: code_review.md (hunk 5 above)
Comment (justinabrahms) [inline code_review.md]: Updated to imply "if it's a partial review, let them know"

File: code_review.md (hunk 6 above)
Comment (justinabrahms) [inline code_review.md]: Done.

File: code_review.md (hunk 7 above)
Comment (justinabrahms) [inline code_review.md]: Done.

Comment (pdpinch) [inline code_review.md]: :shipit:

**Comment by pdpinch** [comment]
:+1:

**Pull request merged**